sql/sql_plugin.cc

```diff
@@ -1581,25 +1584,45 @@ bool update_persisted_plugin_sysvars(const char *name) {
   @param argv actual arguments, propagated to the plugin
   @return Operation outcome, false means no errors
  */
+#ifdef WITH_WSREP
+static bool plugin_register_builtin_and_init_core_se_internal(
+    int *argc, char **argv, st_mysql_plugin **plugins, bool keyring = false) {
+#else
 bool plugin_register_builtin_and_init_core_se(int *argc, char **argv) {
+#endif
   bool mandatory = true;
   DBUG_TRACE;
 
   /* Don't allow initializing twice */
+#ifdef WITH_WSREP
+  /* We need to set 'initialized' flag below, because plugin_foreach_with_mask()
+     uses it in its logic, when key is generated/accessed.
+     However we will call this function again for other plugins than keyring
+     and then assert will fail.
+  */
+  if (keyring) {
+    assert(!initialized);
+  }
+#else
   assert(!initialized);
+#endif
 
   /* Allocate the temporary mem root, will be freed before returning */
   MEM_ROOT tmp_root(key_memory_plugin_init_tmp, 4096);
 
   mysql_rwlock_wrlock(&LOCK_system_variables_hash);
   mysql_mutex_lock(&LOCK_plugin);
+
   initialized = true;
 
   /* First we register the builtin mandatory and optional plugins */
-  for (struct st_mysql_plugin **builtins = mysql_mandatory_plugins;
+  for (struct st_mysql_plugin **builtins = plugins /*mysql_mandatory_plugins*/;
```

Comment:
missing ifdef wsrep - and with this change, the above comment is also incorrect.

Reply:
fixed